### PR TITLE
GH-37673: [C++][Flight] Improve error message if binding port fails

### LINKED ISSUE #37673: [C++][Flight] Improve error message when port cannot be bound with gRPC
### Describe the enhancement requested

Currently, when starting a flight server fails because the call to `arrow::flight::FlightServerBase::Init()` fails internally at the gRPC `BuildAndStart()` step, the error message is rather vague about the reason.

A pretty common error while starting the server is that the supplied port cannot be used; this is indicated by `selected_port` that is populated after `BuildAndStart()`, so the error message could be improved here.

### Component(s)

FlightRPC

### PR DESCRIPTION
**Pull request by romankarlstetter**
See https://github.com/apache/arrow/issues/37673 for more info.
* Closes: #37673

### CODE REVIEW

File: cpp/src/arrow/flight/transport/grpc/grpc_server.cc [hunk 1]
```diff
@@ -598,7 +598,7 @@ class GrpcServerTransport : public internal::ServerTransport {
     } else if (scheme == kSchemeGrpcUnix) {
       std::stringstream address;
       address << "unix:" << uri.path();
-      builder.AddListeningPort(address.str(), ::grpc::InsecureServerCredentials());
```
Comment (lidavidm) [inline cpp/src/arrow/flight/transport/grpc/grpc_server.cc:601]: Looks like this needs to be done above as well for the TLS branch

Comment (lidavidm) [reply cpp/src/arrow/flight/transport/grpc/grpc_server.cc:601]: Plus, for domain sockets, port will still be 0 right?

Comment (romankarlstetter) [reply cpp/src/arrow/flight/transport/grpc/grpc_server.cc:601]: > Looks like this needs to be done above as well for the TLS branch

Hm, I don't really get what you're referring to.





> Plus, for domain sockets, port will still be 0 right?

Yes, that's probably correct (however, I don't have any practical experience with domain sockets in this context).
That would mean that adding the port here does not make sense in the first place and we would need to check below that `scheme != kSchemeGrpcUnix` for issuing that _port bind specific_ error message.

What do you think?

Comment (lidavidm) [reply cpp/src/arrow/flight/transport/grpc/grpc_server.cc:601]: This is an if/else. I think this change should only apply to the TCP socket-based schemes and not the domain socket, so the opposite of the current change.

Comment (romankarlstetter) [reply cpp/src/arrow/flight/transport/grpc/grpc_server.cc]: The more specific message is now only used if the URI scheme is a host/port combination. I hope this logic corresponds to what you had in mind.

File: cpp/src/arrow/flight/transport/grpc/grpc_server.cc [hunk 2]
```diff
@@ -572,7 +572,9 @@ class GrpcServerTransport : public internal::ServerTransport {
 
     const std::string scheme = uri.scheme();
     int port = 0;
-    if (scheme == kSchemeGrpc || scheme == kSchemeGrpcTcp || scheme == kSchemeGrpcTls) {
+    const bool isHostPortScheme =
```
Comment (lidavidm) [inline cpp/src/arrow/flight/transport/grpc/grpc_server.cc:575]: The code style uses snake_case for non-constant variable names

**Review by lidavidm — COMMENTED** [review]
I'm not convinced this is better, e.g. see the failed tests https://github.com/apache/arrow/actions/runs/6403222594/job/17381298479?pr=37689#step:6:3722

```
[ RUN      ] TestTls.DoAction
E1004 08:10:44.675455956   26260 ssl_security_connector.cc:273] Handshaker factory creation failed with TSI_INVALID_ARGUMENT.
E1004 08:10:44.675511559   26260 chttp2_server.cc:1053]      UNKNOWN:Unable to create secure server with credentials of type Ssl {created_time:"2023-10-04T08:10:44.675499712+00:00"}
/arrow/cpp/src/arrow/flight/flight_test.cc:464: Failure
Failed
Expected 'server_->Init(options)' to fail with UnknownError, but got IOError: Server did not start properly. Unable to bind to port 0
```

port 0 doesn't always mean that the port failed to bind, it could mean something else - fundamentally gRPC just does not tell us what failed so I'm not sure we should assume

**Comment by romankarlstetter** [comment]
Ok, I see. So it's probably the best idea to just leave it as is, right? --> I'm closing this.

**Pull request closed without merging**